aiida/orm/data/structure.py

```diff
@@ -873,47 +873,59 @@ def set_pymatgen_structure(self, struct):
             three directions.
         .. note:: Requires the pymatgen module (version >= 3.3.5, usage
             of earlier versions may cause errors).
-        
+
         :raise ValueError: if there are partial occupancies together with spins.
         """
         def build_kind_name(species_and_occu):
             """
-            Build a kind name from a pymatgen Composition, including an
-            additional ordinal if spin is included, e.g. it returns '<element>1'
-            for an atom with spin<0 and '<element>2' for an element with spin>0,
-            ortherwise (no spin) it returns simply '<element>'
-            :param specie: a pymatgen specie
-            :return: a string
+            Build a kind name from a pymatgen Composition, including an additional ordinal if spin is included,
+            e.g. it returns '<specie>1' for an atom with spin < 0 and '<specie>2' for an atom with spin > 0,
+            otherwise (no spin) it returns None
+
+            :param species_and_occu: a pymatgen species and occupations dictionary
+            :return: a string representing the kind name or None
             """
-            has_spin = any([specie.as_dict().get('properties', {}).get('spin', 0) != 0
-                            for specie in species_and_occu.keys()])
+            species = species_and_occu.keys()
+            occupations = species_and_occu.values()
+
+            has_spin = any([specie.as_dict().get('properties', {}).get('spin', 0) != 0 for specie in species])
```

Comment:
`any()` can handle iterators, so you could write `has_spin = any(specie.as_dict() ... for specie in species)`